python/lsst/ip/isr/calibType.py

```diff
@@ -0,0 +1,630 @@
+#
+# LSST Data Management System
+# Copyright 2008-2017 AURA/LSST.
+#
+# This product includes software developed by the
+# LSST Project (http://www.lsst.org/).
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <https://www.lsstcorp.org/LegalNotices/>.
+#
+import abc
+import copy
+import datetime
+import yaml
+
+import lsst.afw.table as afwTable
+from lsst.log import Log
+from lsst.daf.base import PropertyList
+from astropy.table import Table
+
+__all__ = ["IsrCalib", "IsrProvenance"]
+
+
+class IsrCalib(abc.ABC):
+    """Generic calibration type.
+
+    Subclasses must implement the toDict, fromDict, toTable, fromTable
+    methods that allow the calibration information to be converted
+    from dictionaries and afw tables.  This will allow the calibration
+    to be persisted using the base class read/write methods.
+
+    The validate method is intended to provide a common way to check
+    that the calibration is valid (internally consistent) and
+    appropriate (usable with the intended data).  The apply method is
+    intended to allow the calibration to be applied in a consistent
+    manner.
+
+    Parameters
+    ----------
+    detectorName : `str`, optional
+        Name of the detector this calibration is for.
+    detectorSerial : `str`, optional
+        Identifier for the detector.
+    detector : `lsst.afw.cameraGeom.Detector`, optional
+        Detector to extract metadata from.
+    log : `lsst.log.Log`, optional
+        Log for messages.
+
+    """
+    _OBSTYPE = 'generic'
+    _SCHEMA = 'NO SCHEMA'
+    _VERSION = 0
+
+    def __init__(self, detectorName=None, detectorSerial=None, detector=None, log=None, **kwargs):
+        self._detectorName = detectorName
+        self._detectorSerial = detectorSerial
+        self._metadata = PropertyList()
+
+        # Ensure that we have the obs type required by calibration ingest
```

Comment:
This duplicates the code in setMetadata below so I think it might be a good idea to unify this.
For example calling `self.setMetadata(PropertyList())` here.